gpii/node_modules/flowManager/src/MatchMaking.js

```diff
@@ -59,16 +59,22 @@
             fluid.log("gpii.flowManager.getPreferences called - fetching preferences from URL " + prefsServerDataSource.options.prefsServerURL);
         }
 
-        var noUserPrefs = {};
-        var promise = gpiiKey === "noUser" ? fluid.promise().resolve(noUserPrefs) : prefsServerDataSource.get(gpiiKey);
+        var noPrefs = {};
+        var promise = gpiiKey === "noUser" ? fluid.promise().resolve(noPrefs) : prefsServerDataSource.get(gpiiKey);
         promise.then(function (data) {
             that.events.onPreferences.fire(data.preferences || data);
         }, function (err) {
-            that.events.onError.fire({
-                isError: true,
-                message: "Error when retrieving preferences: " + err.message,
-                statusCode: 404
-            });
+            var nonExistentGPiiKeyMsg = "GPII key \"" + gpiiKey + "\" does not exist";
+            // GPII-3721: Return an empty payload for nonexistent GPII keys
+            if (err.message.includes(nonExistentGPiiKeyMsg)) {
```

Comment:
Looking for a substring within a user message isn't a reliable way of implementing core infrastructure. We need to improve the preferences server so that it supplies in the rejection payload an error field with a statically determinable error code, of the kind that we get from HTTP or FS errors, e.g. something like "GPII_ERR_NO_SAFE"

Reply:
Please add a comment referencing GPII-3721 to this site so we can find it easily again when we come to implement the proper grant-sensitive policy, so that we can easily search for the code sites that need to be fixed